lib/lf/uscalfe/fe_space_hp.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * @file
+ * @brief Lagrangian finite elements of arbitrary polynomials degree
+ * @author Tobias Rohner
+ * @date   May 20202
+ * @copyright MIT License
+ */
+
+#ifndef LF_USCALFE_FE_SPACE_HP_H_
+#define LF_USCALFE_FE_SPACE_HP_H_
+
+#include <lf/mesh/utils/all_codim_mesh_data_set.h>
+#include "hp_fe.h"
+#include "scalar_fe_space.h"
+
+namespace lf::uscalfe {
+
+/**
+ * @brief Lagrangian Finite Element Space of arbitrary degree
+ */
+template <typename SCALAR>
+class FeSpaceHP : public ScalarFESpace<SCALAR> {
+ public:
+  using Scalar = SCALAR;
+
+  FeSpaceHP() = delete;
+  FeSpaceHP(const FeSpaceHP &) = delete;
+  FeSpaceHP(FeSpaceHP &&) noexcept = default;
+  FeSpaceHP &operator=(const FeSpaceHP &) = delete;
+  FeSpaceHP &operator=(FeSpaceHP &&) noexcept = default;
+
+  /**
+   * @brief Constructor: Sets up the dof handler
+   * @param mesh_p A shared pointer to the underlying mesh (immutable)
+   * @param N The polynomial degree of the Finite Element Space
+   */
+  explicit FeSpaceHP(std::shared_ptr<const lf::mesh::Mesh> mesh_p, unsigned N)
+      : ScalarFESpace<SCALAR>(std::move(mesh_p)),
+        ref_el_(mesh_p, nullptr),
+        degree_(N) {
+    init();
+  }
+
+  using ScalarFESpace<SCALAR>::Mesh;
+
+  /** @brief access to associated local-to-global map
+   * @return a reference to the lf::assemble::DofHandler object (immutable)
+   */
+  [[nodiscard]] const lf::assemble::DofHandler &LocGlobMap() const override {
+    LF_VERIFY_MSG(Mesh() != nullptr, "No valid FE space object: no mesh");
+    LF_VERIFY_MSG(dofh_p_ != nullptr,
+                  "No valid FE space object: no dof handler");
+    return *dofh_p_;
+  }
+
+  /** @brief access to shape function layout for cells
+   * @copydoc SclarFESpace::ShapeFunctionLayout(const lf::mesh::Entity&)
+   */
+  [[nodiscard]] std::shared_ptr<const ScalarReferenceFiniteElement<SCALAR>>
+  ShapeFunctionLayout(const lf::mesh::Entity &entity) const override {
+    return ref_el_(entity);
+  }
+
+  /** @brief number of _interior_ shape functions associated to entities of
+   * various types
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(
+      const lf::mesh::Entity &entity) const override {
+    return ShapeFunctionLayout(entity)->NumRefShapeFunctions();
+  }
+
+  ~FeSpaceHP() override = default;
+
+ private:
+  // R.H.: Should this really be shared pointers ?
```

Comment:
Should this really be shared pointers? Why not `unique_ptr`?

Reply:
The interface of a `ScalarFESpace` needs a shared pointer to be returned when calling `ShapeFunctionLayout`. Therefore we have no other choice than using shared pointers here. Except of course we would change the interface of the FE spaces to return a reference or raw pointer.

Reply:
This is a very tricky question. So far `UniformScalarFESpace::ShapeFunctionLayout()` has returned a `shared_ptr`. This was of course not a performance problem because that method was not called often. But `ScalarFESpace::ShapeFunctionLayout` can be called for every element in the mesh so that the performance penalty can become relevant.

Currently, the implementation of `FESpaceHP` initializes all reference finite elements and stores them in an `AllCodimMeshDataSet`. Therefore it would be best to just return raw pointers to the underlying `ScalarReferenceFiniteElement`. I.e. the `FESpaceHP` instance owns all FiniteElement instances and determines their lifetime so it only returns a raw pointer.

However, I think that the current approach is slower than the alternative approach of constructing the `FeHP` elements everytime on the fly because
- the construction of a `FeHP` instance is not very expensive
- Nowadays Most CPU Architectures are memory bound, i.e. the bottleneck often lies in the bus speed with which memory is connected to the cpu. Storing a large number of `FeHP` objects into memory and retrieving them again strains exactly the memory bus. Also cache locality is typically much worse because large amounts of memory have to be swapped in and out of the caches.
When we want to construct `FeHP` objects on the fly, **returning a `std::unique_ptr` would indeed be the best solution.**

**My suggestion is to change the interface of `ScalarFESpace` so that it returns a `std::unique_ptr` and construct the `FeHP` objects on-the-fly.**